### PR TITLE
AEA-1082 Feature/dialogues offload

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

handling for dialogues in terminal state

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/protocols/dialogue/base.py [hunk 1]
```diff
@@ -1026,6 +1040,46 @@ def __init__(self, dialogues: "Dialogues") -> None:
...
+        """Get all dialogues in terminal state."""
+        return list(
+            filter(
+                None,
+                [
+                    self._dialogues_by_dialogue_label.get(i)
+                    for i in self._terminal_state_dialogues_labels
+                ],
+            )
+        )
```
Comment (DavidMinarsch) [inline aea/protocols/dialogue/base.py:1056]: But this means we need to keep terminal state dialogue labels..which can grow indefinitely.

Comment (DavidMinarsch) [reply aea/protocols/dialogue/base.py:1056]: Ok no, it will always be empty if not kept, I see

File: aea/configurations/base.py [hunk 2]
```diff
@@ -2099,17 +2101,30 @@ def json(self) -> Dict:
                 },
             }
         )
+        if self.keep_terminal_state_dialogues is not None:
+            result["keep_terminal_state_dialogues"] = self.keep_terminal_state_dialogues
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:2105]: see above; prefer to have a default as we role this in as a new feature now; not an optional one

File: aea/protocols/dialogue/base.py [hunk 3]
```diff
@@ -1026,6 +1051,46 @@ def __init__(self, dialogues: "Dialogues") -> None:
...
+            filter(
+                None,
+                [self._dialogues_by_dialogue_label.get(i) for i in active_dialogues],
+            )
+        )
+
+    @property
+    def is_terminal_dialogues_kept(self) -> bool:
+        """Return True if dialogues should stay after terminal state."""
+        return self._dialogues.is_keep_dialogues_in_terminal_state
```
[the remark this answers is not in the document]
Comment (DavidMinarsch) [reply aea/protocols/dialogue/base.py:1086]: both are ok though.

File: aea/protocols/dialogue/base.py [hunk 4]
```diff
@@ -1026,6 +1051,46 @@ def __init__(self, dialogues: "Dialogues") -> None:
...
+
+    def dialogue_terminal_state_callback(self, dialogue: "Dialogue") -> None:
+        """Method to be called on dialogue terminal state reached."""
+        if self.is_terminal_dialogues_kept:
+            self._terminal_state_dialogues_labels.add(dialogue.dialogue_label)
+        else:
+            self.remove(dialogue.dialogue_label)
```
Comment (DavidMinarsch) [inline aea/protocols/dialogue/base.py:1093]: remove currently does not remove dialogue label from _incomplete_to_complete_dialogue_labels, please fix

File: aea/protocols/dialogue/base.py [hunk 5]
```diff
@@ -1136,53 +1202,88 @@ def _get_collection_name(self) -> Optional[str]:
...
+    ) -> Tuple[Optional[SyncCollection], Optional[SyncCollection]]:
+        """Get active state dialogue collection and terminal collection."""
+        col_name = self._get_collection_name()
+        if not col_name:
+            return (None, None)  # pragma: nocover
+        termina_col_name = f"{col_name}_terminal"
+
+        collection = self._get_collection_instance(col_name)
+        terminal_collection = self._get_collection_instance(termina_col_name)
+        if not collection or not terminal_collection:
```
Comment (DavidMinarsch) [inline aea/protocols/dialogue/base.py:1224]: Not sure why we return like this if either of them is missing...

File: aea/protocols/dialogue/base.py [hunk 6]
```diff
@@ -1212,26 +1313,34 @@ def teardown(self) -> None:
 class Dialogues:
     """The dialogues class keeps track of all dialogues for an agent."""
 
+    _keep_terminal_state_dialogues = True
```
Comment (DavidMinarsch) [inline aea/protocols/dialogue/base.py:1316]: Default should be `False` in my opinion

File: aea/protocols/generator/base.py [hunk 7]
```diff
@@ -1114,6 +1114,12 @@ def _dialogue_class_str(self) -> str:
         cls_str += self.indent + "{" + end_states_str + "}"
         cls_str += self.indent + ")\n\n"
 
+        if self.protocol_specification.keep_terminal_state_dialogues is not None:
```
Comment (DavidMinarsch) [inline aea/protocols/generator/base.py:1117]: see above

File: tests/data/generator/t_protocol/dialogues.py [hunk 8]
```diff
@@ -127,6 +127,8 @@ class TProtocolDialogues(Dialogues, ABC):
         }
     )
 
+    _keep_terminal_state_dialogues = True
```
Comment (DavidMinarsch) [inline tests/data/generator/t_protocol/dialogues.py:130]: why?; I thought we don't modify generated protocols

File: tests/data/generator/t_protocol_no_ct/protocol.yaml [hunk 9]
```diff
@@ -11,7 +11,7 @@ fingerprint:
   message.py: QmTVfQQG8fjXi28qUQgVW3aaaMeam4mJthbjNDT54Jnctc
   serialization.py: QmPX4KzaEfK9JwbbHAmmB1rZp1guJTW5PgPan7ZGAp36DH
   t_protocol_no_ct.proto: Qmc8KkKnWZ9utBxrbEyWhVDRdut87DkFvmHP3SYUg4J3EU
```
Comment (DavidMinarsch) [inline tests/data/generator/t_protocol_no_ct/protocol.yaml:13]: You're using wrong protoc version:
```
protoc --version
libprotoc 3.11.4
```

File: aea/configurations/base.py [hunk 10]
```diff
@@ -2048,6 +2048,7 @@ def __init__(
         license_: str = "",
         aea_version: str = "",
         description: str = "",
+        keep_terminal_state_dialogues: Optional[bool] = None,
```
[the remark this answers is not in the document]
Comment (DavidMinarsch) [reply aea/configurations/base.py:2051]: Needs to go to dialogue section of protocol spec

[the remark this answers is not in the document]
Comment (DavidMinarsch) [reply aea/configurations/base.py:2051]: Protocols:
- contract_api: `keep_terminal_state_dialogues: false`
- default: `keep_terminal_state_dialogues: true`
- fipa: `keep_terminal_state_dialogues: true`
- gym: `keep_terminal_state_dialogues: false`
- http: `keep_terminal_state_dialogues: false`
- ledger_api: `keep_terminal_state_dialogues: false`
- ml_trade: `keep_terminal_state_dialogues: true`
- oef_search: `keep_terminal_state_dialogues: false`
- register: `keep_terminal_state_dialogues: true`
- signing: `keep_terminal_state_dialogues: false`
- state_update: `keep_terminal_state_dialogues: false`
- tac: `keep_terminal_state_dialogues: true`

File: aea/skills/base.py [hunk 11]
```diff
@@ -183,7 +183,7 @@ def outbox(self) -> OutBox:
     @property
     def storage(self) -> Optional[Storage]:
         """Get optional storage for agent."""
-        return self._get_agent_context().storage
+        return self._get_agent_context().storage  # pragma: nocover
```
Comment (DavidMinarsch) [inline aea/skills/base.py:186]: Add tests for these

File: tests/data/generator/t_protocol_no_ct/protocol.yaml (hunk 9 above)
Comment (DavidMinarsch) [reply tests/data/generator/t_protocol_no_ct/protocol.yaml:13]: `python scripts/generate_all_protocols.py`

File: aea/protocols/generator/base.py (hunk 7 above)
Comment (solarw) [reply aea/protocols/generator/base.py]: done

File: tests/data/generator/t_protocol_no_ct/protocol.yaml (hunk 9 above)
Comment (solarw) [reply tests/data/generator/t_protocol_no_ct/protocol.yaml:13]: done

File: aea/skills/base.py (hunk 11 above)
Comment (solarw) [reply aea/skills/base.py]: done

File: aea/protocols/dialogue/base.py (hunk 6 above)
Comment (solarw) [reply aea/protocols/dialogue/base.py]: done

File: aea/protocols/dialogue/base.py (hunk 5 above)
Comment (solarw) [reply aea/protocols/dialogue/base.py:1224]: discussed

File: aea/configurations/base.py [hunk 12]
```diff
@@ -2099,17 +2101,30 @@ def json(self) -> Dict:
...
+        if self.keep_terminal_state_dialogues is not None:
+            result["keep_terminal_state_dialogues"] = self.keep_terminal_state_dialogues
+        return result
 
     @classmethod
     def from_json(cls, obj: Dict):
         """Initialize from a JSON object."""
+        kwargs = {}
+
+        if obj.get("keep_terminal_state_dialogues") is not None:
```
[the remark this answers is not in the document]
Comment (solarw) [reply aea/configurations/base.py]: done

File: aea/configurations/base.py (hunk 2 above)
Comment (solarw) [reply aea/configurations/base.py]: done

File: aea/configurations/base.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (solarw) [reply aea/configurations/base.py]: done

File: aea/protocols/dialogue/base.py [hunk 13]
```diff
@@ -1278,6 +1387,11 @@ def __init__(
         )
         self._role_from_first_message = role_from_first_message
 
+    @property
+    def is_keep_dialogues_in_terminal_state(self) -> bool:
+        """Is requrired to keep dialogues in terminal state."""
+        return self._keep_terminal_state_dialogues  # pragma: nocover
```
Comment (DavidMinarsch) [inline aea/protocols/dialogue/base.py:1393]: add a test

File: aea/skills/base.py [hunk 14]
```diff
@@ -577,13 +577,37 @@ def parse_module(  # pylint: disable=arguments-differ
...
+            self._keep_terminal_state_dialogues = keep_terminal_state_dialogues
+
+    def setup(self) -> None:  # pragma: nocover
         """Set the class up."""
         super_obj = super()
         if hasattr(super_obj, "setup"):
             super_obj.setup()  # type: ignore  # pylint: disable=no-member
 
-    def teardown(self) -> None:
+    def teardown(self) -> None:  # pragma: nocover
```
Comment (DavidMinarsch) [inline aea/skills/base.py:610]: add test

File: aea/protocols/dialogue/base.py (hunk 4 above)
Comment (solarw) [reply aea/protocols/dialogue/base.py:1093]: done

File: aea/protocols/dialogue/base.py (hunk 13 above)
Comment (solarw) [reply aea/protocols/dialogue/base.py]: done already

File: aea/skills/base.py (hunk 14 above)
Comment (solarw) [reply aea/skills/base.py]: done already

File: aea/protocols/dialogue/base.py [hunk 15]
```diff
@@ -1136,53 +1202,88 @@ def _get_collection_name(self) -> Optional[str]:
...
+    @staticmethod
+    def _dump_dialogues(
+        dialogues: Iterable[Dialogue], collection: SyncCollection
+    ) -> None:
+        """Dump dialogues to collection."""
+        for dialogue in dialogues:
+            collection.put(str(dialogue.dialogue_label), dialogue.json())
+
+    def _load(self) -> None:
+        """Dump dialogues and incomplete dialogues from the generic storage."""
```
Comment (marcofavorito) [inline aea/protocols/dialogue/base.py:1273]: "Load dialogues", not "Dump dialogues"

Comment (marcofavorito) [reply aea/protocols/dialogue/base.py:1273]:
```suggestion
        """Load dialogues and incomplete dialogues from the generic storage."""
```

File: aea/protocols/dialogue/base.py [hunk 16]
```diff
@@ -1212,26 +1313,34 @@ def teardown(self) -> None:
...
         :return: None
         """
 
         self._dialogues_storage = PersistDialoguesStorage(self)
         self._self_address = self_address
         self._dialogue_stats = DialogueStats(end_states)
 
+        if keep_terminal_state_dialogues is not None:
+            self._keep_terminal_state_dialogues = keep_terminal_state_dialogues
+
```
Comment (marcofavorito) [inline aea/protocols/dialogue/base.py:1343]: else?

Comment (DavidMinarsch) [reply aea/protocols/dialogue/base.py:1343]: It's updating the class variable which is set by default

**Review by marcofavorito — APPROVED** [review]
LGTM

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**